### DISCUSSION TITLE
Setting up Terraform

### DISCUSSION THREAD

**Discussion by azdanov** [LazyVim/LazyVim#982] [Q&A] [answered]
Hey, I'm a beginner with vim (1 week). How should I set up [Terraform ](https://terraform.io) (lsp, autocomplete)?
I understood it's by creating a custom plugin. But which packages and settings should I use?

I'd appreciate if someone has any ideas 🙂.

**Comment by robsonpeixoto** [comment, marked as answer]
I'm using this config: `$HOME/.config/nvim/lua/plugins/terraform.lua`

```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      if type(opts.ensure_installed) == "table" then
        vim.list_extend(opts.ensure_installed, { "terraform", "hcl" })
      end
    end,
  },

  {
    "neovim/nvim-lspconfig",
    ---@class PluginLspOpts
    opts = {
      ---@type lspconfig.options
      servers = {
        terraformls = {},
      },
    },
  },
}
```

**Reply by azdanov** [reply]
Thanks, it works!

For some reason formatting doesn't work. Is there any additional config that I would need? I thought terraformls would be enough.

**Reply by robsonpeixoto** [reply]
just add: 
```lua
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local null_ls = require("null-ls")
      vim.list_extend(opts.sources, {
        null_ls.builtins.formatting.terraform_fmt,
        null_ls.builtins.formatting.terrafmt, -- markdown
        null_ls.builtins.diagnostics.terraform_validate,
      })
    end,
  },

```

**Reply by azdanov** [reply]
Awesome, thanks!

**Reply by azdanov** [reply]
@robsonpeixoto Thank again for help! Maybe you have some clues as to why tflint only works on terraform files that are in the root, but it stops working inside directories?
![image](https://github.com/LazyVim/LazyVim/assets/6123841/bb60880a-3195-4bea-a873-ff25228c05c5)

**Reply by robsonpeixoto** [reply]
Good question! There is a typo(`a mi`) on the code, is not because of that ?

What's happen when you call tflint directly from command line ?

**Reply by azdanov** [reply]
It was just me trying to show that lsp is working with that typo 😄

As for tflint cli, it works whenever I call it in the directory that has `.tflint.hcl`. So maybe when I open the subdirectory, then tflint doesn't know how to find the config file in root directory?

**Reply by azdanov** [reply]
Something like this, where tflint inside nvim doesn't find the config.
![image](https://github.com/LazyVim/LazyVim/assets/6123841/021ac90e-24f0-47e0-b92b-ad0b0efdc1e0)

**Reply by azdanov** [reply]
https://github.com/terraform-linters/tflint/blob/master/docs/user-guide/config.md#configuring-tflint

I guess I'd need either to add `--recursive` or to specify `--config "$(pwd)/.tflint.hcl"` somehow.

**Reply by robsonpeixoto** [reply]
How did you configured the tflint ?

**Reply by azdanov** [reply]
```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      if type(opts.ensure_installed) == "table" then
        vim.list_extend(opts.ensure_installed, { "terraform", "hcl" })
      end
    end,
  },
  {
    "neovim/nvim-lspconfig",
    ---@class PluginLspOpts
    opts = {
      ---@type lspconfig.options
      servers = {
        terraformls = {},
        tflint = {},
      },
    },
  },
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local null_ls = require("null-ls")
      vim.list_extend(opts.sources, {
        null_ls.builtins.formatting.terraform_fmt,
        null_ls.builtins.formatting.terrafmt, -- markdown
        null_ls.builtins.diagnostics.terraform_validate,
      })
    end,
  },
}
```

**Reply by azdanov** [reply]
If I put the `.tflint.hcl` in my $HOME, then it works.

**Reply by azdanov** [reply]
Okay, it started working out of nowhere. I'm not sure what was the issue 😵
@robsonpeixoto Thank you for your time. I guess I'll need to learn more about neovim and lsp ecosystem, it's so complex.
